### PR TITLE
rbd: demote/promote all mirrored images in a pool

### PR DESCRIPTION
**Pull request by dillaman**

### CODE REVIEW

**Comment by dillaman** [comment]
Currently fails the rbd-mirror workunit test

File: src/librbd/CMakeLists.txt [hunk 1]
```diff
@@ -91,8 +92,9 @@ set(librbd_internal_srcs
   operation/SnapshotUnprotectRequest.cc
   operation/SnapshotLimitRequest.cc
   operation/TrimRequest.cc
-  watcher/RewatchRequest.cc
+  watcher/Notifier.cc
   watcher/Types.cc
+  watcher/RewatchRequest.cc
```
Comment (trociny) [inline src/librbd/CMakeLists.txt]: It should be above Types.cc?

File: src/librbd/api/Image.cc [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+                 << dendl;
+      return r;
+    }
+
+    set<string> image_ids;
+    r = cls_client::get_children(&ioctx, RBD_CHILDREN, parent_spec,
+                                 image_ids);
+    if (r < 0 && r != -ENOENT) {
+      lderr(cct) << "Error reading list of children from pool " << it->second
+      	   << dendl;
```
Comment (trociny) [inline src/librbd/api/Image.cc]: indentation

File: src/librbd/api/Mirror.cc [hunk 3]
```diff
@@ -0,0 +1,887 @@
...
+
+template <typename I>
+int validate_mirroring_enabled(I *ictx) {
+  CephContext *cct = ictx->cct;
+  cls::rbd::MirrorImage mirror_image_internal;
+  int r = cls_client::mirror_image_get(&ictx->md_ctx, ictx->id,
+      &mirror_image_internal);
+  if (r < 0 && r != -ENOENT) {
+    lderr(cct) << "failed to retrieve mirroring state: " << cpp_strerror(r)
+      << dendl;
```
Comment (trociny) [inline src/librbd/api/Mirror.cc]: identation

File: src/librbd/api/Mirror.cc [hunk 4]
```diff
@@ -0,0 +1,887 @@
...
+    return r;
+  } else if (mirror_image_internal.state !=
+               cls::rbd::MIRROR_IMAGE_STATE_ENABLED) {
```
Comment (trociny) [inline src/librbd/api/Mirror.cc]: not sure about indentation here

File: src/librbd/api/Mirror.cc [hunk 5]
```diff
@@ -0,0 +1,887 @@
...
+                       std::set<std::string>& mirror_image_ids) {
+  CephContext *cct = reinterpret_cast<CephContext *>(io_ctx.cct());
+
+  std::string last_read = "";
+  int max_read = 1024;
+  int r;
+  do {
+    std::map<std::string, std::string> mirror_images;
+    r =  cls_client::mirror_image_list(&io_ctx, last_read, max_read,
+                                           &mirror_images);
```
Comment (trociny) [inline src/librbd/api/Mirror.cc]: identation

File: src/librbd/api/Mirror.cc [hunk 6]
```diff
@@ -0,0 +1,887 @@
...
+    if (r < 0) {
+      lderr(cct) << "error listing mirrored image directory: "
+           << cpp_strerror(r) << dendl;
```
Comment (trociny) [inline src/librbd/api/Mirror.cc]: identation

File: src/librbd/api/Mirror.cc [hunk 7]
```diff
@@ -0,0 +1,887 @@
...
+  if (r < 0) {
+    return r;
+  }
+
+  cls::rbd::MirrorImage mirror_image_internal;
+  r = cls_client::mirror_image_get(&ictx->md_ctx, ictx->id,
+      &mirror_image_internal);
+  if (r < 0 && r != -ENOENT) {
+    lderr(cct) << "failed to retrieve mirroring state: " << cpp_strerror(r)
+      << dendl;
```
Comment (trociny) [inline src/librbd/api/Mirror.cc]: identation

File: src/librbd/api/Mirror.cc [hunk 8]
```diff
@@ -0,0 +1,887 @@
...
+    return r;
+  }
+
+  cls::rbd::MirrorImageStatus
+    s(cls::rbd::MIRROR_IMAGE_STATUS_STATE_UNKNOWN, "status not found");
+
+  r = cls_client::mirror_image_status_get(&ictx->md_ctx, info.global_id, &s);
+  if (r < 0 && r != -ENOENT) {
+    lderr(cct) << "failed to retrieve image mirror status: "
+      	 << cpp_strerror(r) << dendl;
```
Comment (trociny) [inline src/librbd/api/Mirror.cc]: identation

File: src/librbd/api/Mirror.cc [hunk 9]
```diff
@@ -0,0 +1,887 @@
...
+                     << cpp_strerror(close_r) << dendl;
+          return close_r;
+        }
+      }
+    }
+  }
+
+  r = cls_client::mirror_mode_set(&io_ctx, next_mirror_mode);
+  if (r < 0) {
+    lderr(cct) << "Failed to set mirror mode: " << cpp_strerror(r) << dendl;
```
Comment (trociny) [inline src/librbd/api/Mirror.cc]: While here, there might be a sense to unify all messages to start lowercase?

File: src/test/rbd_mirror/test_PoolWatcher.cc [hunk 10]
```diff
@@ -68,11 +69,12 @@ TestPoolWatcher() : m_lock("TestPoolWatcherLock"),
...
-      ASSERT_EQ(0, librbd::mirror_mode_set(ioctx, RBD_MIRROR_MODE_POOL));
+      ASSERT_EQ(0, librbd::api::Mirror<>::mode_set(ioctx,
+                                                   RBD_MIRROR_MODE_POOL));
       std::string uuid;
-      ASSERT_EQ(0, librbd::mirror_peer_add(ioctx, &uuid,
-					   peer.cluster_name,
-					   peer.client_name));
+      ASSERT_EQ(0, librbd::api::Mirror<>::peer_add(ioctx, &uuid,
+					           peer.cluster_name,
+					           peer.client_name));
```
Comment (trociny) [inline src/test/rbd_mirror/test_PoolWatcher.cc]: on github indentation looks wrong (needs an additional whitespace)?

File: src/librbd/mirror/DemoteRequest.h [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+   * GET_INFO
+   *    |
+   *    v
+   * ACQUIRE_LOCK * * * *
+   *    |               *
+   *    v               *
+   * DEMOTE             *
+   *    |               *
+   *    v               *
+   * RELEASE_LOCK < * * *
```
Comment (trociny) [inline src/librbd/mirror/DemoteRequest.h]: The error arrow from ACQUIRE_LOCK should go to <finish>?

File: src/librbd/api/Mirror.cc (hunk 4 above)
Comment (dillaman) [inline src/librbd/api/Mirror.cc]: looks good to me -- indented two spaces since it's a carryover from the line above

Comment (trociny) [inline src/librbd/api/Mirror.cc]: ok. Just we have many examples in our code where it starts after the '(' above, i.e.:
```
if (x
    y)
```
And this is how Emacs indents to me.

Comment (dillaman) [inline src/librbd/api/Mirror.cc]: I always code with wrapped conditional clauses indented to show that they are grouped w/ the previous line. I.e.:

```
if (something_very_long !=
      something_else &&
    some_other_very_long_conditional_clause) {
  ...
}
```

vs

```
if (something_very_long !=
    something_else &&
    some_other_very_long_conditional_clause) {
  ...
}
```

File: src/test/rbd_mirror/test_PoolWatcher.cc (hunk 10 above)
Comment (dillaman) [inline src/test/rbd_mirror/test_PoolWatcher.cc]: Looks correct in an editor -- just github not liking mixing tabs and spaces throughout the codebase

**Comment by dillaman** [comment]
retest this please

**Comment by dillaman** [comment]
retest this please

**Pull request merged**